### ISSUE TITLE
Mediated transfers not working with low but valid settle/reveal timeouts

### ISSUE THREAD

**Issue by andrevmatos** [raiden-network/raiden#4998]
## Problem Definition

Testing PFS mediated transfers from Light Client, with low timeouts, break mediated transfers. The confirmation blocks was reduced to 2 on all clients to avoid `2 * confirmation_blocks >= reveal_timeout`, but even then, mediated transfers simply aren't forwarded. 
This was performed with 2 fresh raiden accounts on the full nodes, on Goerli, loaded with ETH and TTT tokens. Direct channels between `0xE090 (LC) -> 0xaAaA ->0x0081`. Trying first a direct transfer between LC and 0xaAaA worked, but trying from the LC to 0x0081 wasn't forwarded and then expired.
I can see `ActionInitMediator` log on the mediator, with `Processed` being sent back to the LC, but no error message is seen and no message is sent to target. Target has no relevant logs.

The changed defaults/constants are:
```
DEFAULT_REVEAL_TIMEOUT = 10
DEFAULT_SETTLE_TIMEOUT = 20
DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS = 2
DEFAULT_CHANNEL_SYNC_TIMEOUT = 2
```

- 0xaAaA logs (mediator): [raiden-debug_2019-09-28T22:34:44.196414.log](https://github.com/raiden-network/raiden/files/3666100/raiden-debug_2019-09-28T22.34.44.196414.log)
- 0x0081 logs (target): [raiden-debug_2019-09-28T22:34:42.676791.log](https://github.com/raiden-network/raiden/files/3666101/raiden-debug_2019-09-28T22.34.42.676791.log)
- Databases: [raiden_mediated_fail.tar.gz](https://github.com/raiden-network/raiden/files/3666102/raiden_mediated_fail.tar.gz)

### System Description

Raiden on docker, but not the official images but instead `develop` with changed `settings.py` as per above:
```
{
  "raiden": "0.100.5a1.dev456+g684cef6f5.d20190928",
  "raiden_db_version": 23,
  "python_implementation": "CPython",
  "python_version": "3.7.4",
  "system": "Linux 64bit_ELF 4.19.75-1-lts",
  "architecture": "x86_64",
  "distribution": "source"
}
```

**Comment by rakanalh** [comment]
I am seeing the exact same issue in my PR: for #4999 when decreasing the reveal timeout * settle timeout. I am currently investigating this one.

**Comment by rakanalh** [comment]
Investigated the issue and figured out the following:

Setting lower reveal timeouts(say 20) on a node level means that this node will use the reveal timeout to calculate the lock expiry for any outgoing transfer. However, for the receiving node, the default value is used (50 in settings.py). This means that as soon as the receiving node gets the transfer, it will not send a `SecretRequest` because it deems that it `not is_safe_to_wait` with the remaining blocks given it's reveal timeout value in that node.

Discussed this with @hackaugusto and this could get quite complicated quite fast:
- Having lower reveal timeouts means that there is a higher risk in the secret registration transaction not being mined before the lock expires. 
- For the sender node with the lower reveal timeout: It bares no risk because it doesn't have to register the secret on-chain.
- For the receiver node with the lower reveal timeout: Bares all the risk because it is the one that has to manage to register the secret on-chain before the lock expires.

Possibilities to resolve the above issue:
1. Before node A creates the channel, it communicates its lower reveal timeout to it's partner node B.  Node B in this case would only use A's reveal timeout to calculate the expiration for transfers going to A but not coming from A because B doesn't want to bare any risks when it comes to received transfers. This complicates the implementation we have right now.
2. Register the reveal timeout value on-chain in the channel state with the settle timeout. That basically means that storage is not optimized on-chain given that we had to store an additional value. That being said, we also somehow enforce A's chosen reveal timeout on B when A creates a channel with B which as a result means that B needs to bare some risk in accepting to work in channels with lower reveal timeouts.

**Comment by andrevmatos** [comment]
@rakanalh I'm not sure this was specifically the issue that happened here. The mediating and target nodes were configured with same timeouts, and also the receiver actually never got the message, as you can see in the mediator logs, `ActionInitMediator` is fired but no `LockedTransfer` is ever sent.
Funny enough, I hit this issue again today, with default confirmation blocks (`5`), `settle_timeout=100` and `reveal_timeout=19` on all mediating nodes. The transfer got stuck on the first hop, it was accepted, `ActionInitMediator` was fired but no locked transfer was sent, BUT upon restarting the first mediator (which held the transfer), it detected the pending mediated transfer, sent it, and the protocol completed!

**Comment by rakanalh** [comment]
One thing you can do is run `python tools/debugging/replay_wal.py` and see what events were emitted when the `ActionInitMediator` is received. I suspect that it has something to do with `is_safe_to_wait`. You can PDB in the state machine and follow the logic to figure out what happens. If you're able to do this, let us know what you find please.

**Comment by konradkonrad** [comment]
So the issue seems to arise here:
```
> ActionInitMediator
	sender='0xe0902284c85a9a03daa3b5ab032e238cc05cff9a'
	balance_proof=[...]
	from_hop=[...]
	route_states=[...]
	from_transfer=LockedTransferSignedState(payment_identifier=18290067926980017390, token=..., lock=HashTimeLockState(amount=..., expiration=1379501, secrethash=<manually_hex_encoded:>'0x14c5d23c09ddf9eeaf36654fcaf152a2767904263f4734528016291e5b6d50ab', message_identifier=1569710796089, routes=[...])
< SendProcessed
	recipient='0xe0902284c85a9a03daa3b5ab032e238cc05cff9a'
	canonical_identifier=[...]
	message_identifier=1569710796089
	queue_identifier=[...]
> ReceiveDelivered
	sender='0xe0902284c85a9a03daa3b5ab032e238cc05cff9a'
	message_identifier=1569710796089
> ReceiveDelivered
	sender='0xe0902284c85a9a03daa3b5ab032e238cc05cff9a'
	message_identifier=1569710796089
> Block
	block_number=1379481
	gas_limit=8000000
	block_hash='0x4c5edebab3b71a0e2147a8aefa438d95fb10ca3e4ed9a9eab3eb08956b4d62a3'
< EventUnlockClaimFailed
	identifier=18290067926980017390
	secrethash='0x14c5d23c09ddf9eeaf36654fcaf152a2767904263f4734528016291e5b6d50ab'
	reason='lock expired'
```

Even though there are 20 Blocks left, the mediator fires `events_for_expired_pairs`.

**Comment by rakanalh** [comment]
Ok, since you're talking about expirations, here's what i think:

https://github.com/raiden-network/raiden/blob/af1050b5bd1872517d4c70e06d9d532cbe69d42f/raiden/transfer/channel.py#L238-L241

The above makes sure that the `lock_timeout` is between `reveal_timeout` and `settle_timeout`. I would check the logs to see what `lock_timeout` value was passed to the transfer API call. If it was None, then the default will be calculated:
```
return BlockExpiration(block_number + reveal_timeout * 2)
```

Make sure that in the above call the condition doesn't fail. If it does, then that's where the problem lies. In this case, we should check why our calculated value wasn't passing our criterias.

**Comment by rakanalh** [comment]
I've had an issue in the scenarios that prevented mediators from sending the transfer forward because the lock timeout was invalid. The solution we added into the scenario was to send the `lock_timeout` param to `30` provided that the scenario used `20` for reveal_timeout and `40` for settle_timeout.

**Comment by konradkonrad** [comment]
The mediator seems to fail this transfer due to incompatible timeout settings:
https://github.com/raiden-network/raiden/blob/2f6703e154438bd44ad2bfd6fb734d58ce6df856/raiden/transfer/channel.py#L238-L241
fails because
```
lock_timeout = 21
channel_state.settle_timeout = 20
```
Therefore
https://github.com/raiden-network/raiden/blob/2f6703e154438bd44ad2bfd6fb734d58ce6df856/raiden/transfer/channel.py#L207-L209
fails.
Therefore
https://github.com/raiden-network/raiden/blob/2f6703e154438bd44ad2bfd6fb734d58ce6df856/raiden/transfer/mediated_transfer/mediator.py#L417-L422
does not emit any events. This leads to an attempted backward_transfer
https://github.com/raiden-network/raiden/blob/2f6703e154438bd44ad2bfd6fb734d58ce6df856/raiden/transfer/mediated_transfer/mediator.py#L1081-L1084
that fails for the same reason.

**Comment by konradkonrad** [comment]
So the explanation for the failure is as follows:
- By default, the `lock_expiration` on sender side is configured as `current_block + 2 * reveal_timeout`, in your case `current_initiator_block + 20`.
- It seems, that the mediator, at the time of receiving the transfer, is 1 block behind. So it calculates the `lock_timeout` from the `expiration` as `21`.

A solution, if you need to use such low timeouts, is to employ manual `lock_timeouts` via the API, see 

https://github.com/raiden-network/raiden/blob/2f6703e154438bd44ad2bfd6fb734d58ce6df856/raiden/tests/scenarios/bf1_basic_functionality.yaml#L40-L41

and 

https://github.com/raiden-network/raiden/blob/2f6703e154438bd44ad2bfd6fb734d58ce6df856/raiden/tests/scenarios/bf1_basic_functionality.yaml#L103

Or make sure, there is a reasonable margin for `2 * reveal_timeout - settle_timeout > your_expected_block_synchronization_difference_between_clients`.

**Comment by andrevmatos** [comment]
@konradkonrad I'd say this is a workaround, not a fix. These checks should be done before accepting the transfer, and proper error logged/reported. Once `ActionInitMediator` is emitted, it should be acted upon. If  `Processed` was sent, it's expected the mediator will forward the transfer. If the check dropped the transfer before sending this, retry mechanism would send it again later once the mediator catched upon the right block height, and transfer would be performed. It having accepted the transfer, emitted the Action, even replied with `Processed`, makes it unaceptable for a later check to silently drop the transfer, IMO. So I'd say it's not fixed.

**Comment by konradkonrad** [comment]
I can follow your argument, and I think it is valid to re-open this. However, this will need some discussion, as to what properties of a `LockedTransfer` define the pre-`Processed` acceptance criteria.

**Comment by rakanalh** [comment]
> These checks should be done before accepting the transfer, and proper error logged/reported.

This is already done. Candidate routes are filtered according to whether a certain channel is usable or not:

https://github.com/raiden-network/raiden/blob/df45b930714456aaed3fa2910850b7f08b017fa7/raiden/transfer/mediated_transfer/initiator.py#L265-L273

Which means that the initiator will just filter out the route if the lock_timeout value is invalid (not within the range `reveal_timeout < {lock_timeout} <= settle timeout`.
With that beeing said, this can only be done with the "next hop" route rather than every route. So you are guaranteed that at some point the transfer will fail if each node uses a custom reveal/settle timeouts.

> Once ActionInitMediator is emitted, it should be acted upon. If Processed was sent, it's expected the mediator will forward the transfer.

This is not how it works right now. `Processed` confirms receiving and processing the action but doesn't necessarily mean that the transfer went forward. However, the mediator does **not** drop the transfer but rather, it sets it's `waiting_transfer` property to this transfer. Unfortunately, we currently only retry the `waiting_transfer` if the partner's network state changes. 
Having the partner retry sending the transfer would mean a new `ActionInitMediiator` state change with the same balance proof & same nonce and this would actually fail during validation. This is why we send `Processed` which is to confirm to the partner that we **did** process the transfer. "Couldn't move it forward" is something else that we can probably do something about.

I would say this is something that has to be thought and there's no immediate fix we can do. The recommendation is to use reasonable settle/reveal/lock timeout values that all the nodes in the path can work with.

**Comment by rakanalh** [comment]
The above issue was created to decide how to move forward.
